rmw_fastrtps_cpp/src/functions.cpp

```diff
@@ -1750,10 +1702,9 @@ extern "C"
             {
                 void *data = guard_conditions->guard_conditions[i];
                 GuardCondition *guard_condition = (GuardCondition*)data;
-                if(!guard_condition->getHasTriggered())
-                {
-                    guard_conditions->guard_conditions[i] = 0;
-                }
+                // Guard conditions persist (don't set to null)
+                // We call getHasTriggered to reset the trigger state.
```

Comment:
> Without this we are not able to check that our code is sleeping correctly. Simply because of that I think it is crucial to keep.

That's the debugging argument, which without addressing the fact that we don't have a good way to do the debugging yet is to just reiterate what has already been argued.

My opinion is that we leave it until we need it or we have the tools to responsibly log it, but I can't honestly say that opinion is any more correct than @jacquelinekay's which is that if we're not using it we should get rid of it.

> Second: the only reason why our executor has to refill all the handles every time before wait is because it simply doesn't use the information if the interrupt condition was triggered (in order to use new/changes handles). I think that is a good example why using this information is useful to avoid overhead like this.

I don't understand what you mean here. As I understand it, the reason we have to refill the handles each time is because the rmw layer will replace pointers to the guard conditions with null anytime they're not set after waking up. Also that the set of guard conditions being passed may change from loop to loop (like when nodes are added or removed). So I think the simplest thing to do is to clear the guard conditions and add them all each time regardless of what has been waited on in the past and what was triggered or not last loop. This isn't an expensive set of operations, it's basically a memset to 0 for the waitset storage and then assigning the pointers into the waitset's storage. You might have to resize the waitset storage which is more expensive but hopefully infrequent.

I suppose you could maintain additional state and only selectively "reinsert" pointers into the waitset storage when entering a new loop, but I think that would be more computation than just reassigning them all each time. Also reassigning them all each time couldn't be avoided when you need to resize the waitset or when the set of things being waited on has changed. So I think it would add a lot of complexity for relatively little benefit.